Arrowgene.Ddon.GameServer/Handler/CraftStartEquipGradeUpHandler.cs

```diff
@@ -0,0 +1,210 @@
+            }
+            else
+            {
+                res = CreateEquipPointResponse(equipItemUID, addEquipPoint, currentTotalEquipPoint, goldRequired, IsGreatSuccess, CurrentEquipInfo, canContinue, dummydata);
+            }
+
+            // TODO: Store saved pawn exp
+            S2CCraftCraftExpUpNtc expNtc = new S2CCraftCraftExpUpNtc()
+            {
+                PawnId = request.CraftMainPawnID,
+                AddExp = PawnExp,
+                ExtraBonusExp = 0,
+                TotalExp = PawnExp,
+                CraftRankLimit = 0
+            };
+            client.Send(expNtc);
+
+            return res;
+        }
+
+        private void UpdateCharacterItem(GameClient client, string equipItemUID, Item equipItem, List<CDataCommonU32> gradeuplist, uint charid, uint pawnid, S2CItemUpdateCharacterItemNtc updateCharacterItemNtc, CDataCurrentEquipInfo CurrentEquipInfo)
+        {
+            var (storageType, foundItem) = client.Character.Storage.FindItemByUIdInStorage(ItemManager.EquipmentStorages, equipItemUID);
+            if (foundItem != null)
+            {
+                var (slotno, item, itemnum) = foundItem;
+                CharacterCommon characterCommon = null;
+                if (storageType == StorageType.CharacterEquipment || storageType == StorageType.PawnEquipment)
+                {
+                    CurrentEquipInfo.EquipSlot.EquipSlotNo = EquipManager.DetermineEquipSlot(slotno);
+                    CurrentEquipInfo.EquipSlot.EquipType = EquipManager.GetEquipTypeFromSlotNo(slotno);
+                }
+
+                if (storageType == StorageType.PawnEquipment)
+                {
+                    CurrentEquipInfo.EquipSlot.PawnId = pawnid;
+                    characterCommon = client.Character.Pawns.SingleOrDefault(x => x.PawnId == pawnid);
+                }
+                else if (storageType == StorageType.CharacterEquipment)
+                {
+                    CurrentEquipInfo.EquipSlot.CharacterId = charid;
+                    characterCommon = client.Character;
+                }
+
+                updateCharacterItemNtc.UpdateType = ItemNoticeType.StartEquipGradeUp;
+                updateCharacterItemNtc.UpdateItemList.Add(Server.ItemManager.CreateItemUpdateResult(characterCommon, equipItem, storageType, (byte)slotno, 0, 0));
+
+                _itemManager.UpgradeStorageItem(Server, client, charid, storageType, equipItem, (byte)slotno);
+                updateCharacterItemNtc.UpdateItemList.Add(Server.ItemManager.CreateItemUpdateResult(characterCommon, equipItem, storageType, (byte)slotno, 1, 1));
+                client.Send(updateCharacterItemNtc);
+            }
+            else
+            {
+                Logger.Error($"Item with UID {equipItemUID} not found in {storageType}");
+                throw new Exception($"Item with UID {equipItemUID} not found in {storageType}");
```

Comment:
Sorry, I was not more clear in my previous comment. You should really throw exceptions using `ResponseErrorException`. It allows you to assign an error code which will show up in the client UI. Do this for all the exceptions you just added.

Do something like this (pick a better error code if you can find one).
```c#
ResponseErrorException(ErrorCode.ERROR_CODE_ITEM_INVALID_UID, $"Item with UID {equipItemUID} not found in {storageType}");
```